pipeline_dp/budget_accounting.py

```diff
@@ -100,7 +101,188 @@ def compute_budgets(self):
+            total_delta: delta for the entire pipeline.
+            pld_discretization: `value_discretization_interval` in PLD library.
+                Smaller interval results in better accuracy, but increases running time.
+
+        Raises:
+            ValueError: Arguments are missing or out of range.
+        """
+
+        _validate_epsilon_delta(total_epsilon, total_delta)
+
+        self._total_epsilon = total_epsilon
+        self._total_delta = total_delta
+        self._mechanisms = []
+        self.minimum_noise_std = None
+        self._pld_discretization = pld_discretization
+
+    def request_budget(self, noise_kind: NoiseKind) -> MechanismSpec:
+        """Request a budget.
+
+        Constructs a mechanism spec based on the parameters.
+        Adds the mechanism to the pipeline for future calculation.
+
+        Args:
+            noise_kind: The kind of noise distribution for the mechanism.
+
+        Returns:
+            A "lazy" mechanism spec object that doesn't contain the noise
+            standard deviation until compute_budgets is called.
+        """
+        mechanism_spec = MechanismSpec(noise_kind=noise_kind)
+        mechanism_spec_internal = MechanismSpecInternal(
+            mechanism_spec=mechanism_spec)
+        self._mechanisms.append(mechanism_spec_internal)
+        return mechanism_spec
+
+    def compute_budgets(self):
+        """Computes the budget for the pipeline.
+
+        Composes the mechanisms and adjusts the amount of
+        noise based on given epsilon. Sets the noise for the
+        entire pipeline.
+        """
+        if not self._mechanisms:
+            return
+        minimum_noise_std = self._find_minimum_noise_std()
+        self.minimum_noise_std = minimum_noise_std
+        for mechanism in self._mechanisms:
+            mechanism.mechanism_spec._noise_standard_deviation = minimum_noise_std / mechanism.weight
+
+    def _find_minimum_noise_std(self) -> float:
+        """Find a noise value to the minimum.
+
+        Use binary search to find a minimum noise value that gives a
+        new epsilon close to the given epsilon (within a threshold).
+        By increasing the noise we can decrease the epsilon.
+
+        Returns:
+            The noise value adjusted for the given epsilon.
+        """
+        threshold = 1e-4
+        maximum_noise_std = self._calculate_max_noise_std()
+        low, high = 0, maximum_noise_std
+        while low + threshold < high:
+            mid = (high - low) / 2 + low
+            pld = self._compose_distributions(mid)
+            pld_epsilon = pld.get_epsilon_for_delta(self._total_delta)
+            if pld_epsilon <= self._total_epsilon:
+                high = mid
+            elif pld_epsilon > self._total_epsilon:
+                low = mid
+
+        return high
+
+    def _calculate_max_noise_std(self) -> float:
+        """Calculate maximum noise std for given epsilon.
+
+        Returns:
```

Comment:
It looks like Returns section doesn't provide additional infomation to function description. Please drop it.